core/src/main/java/org/jboss/jca/core/connectionmanager/pool/AbstractPool.java

```diff
@@ -505,55 +505,70 @@ public void flush(boolean kill)
+      synchronized (this)
+      {
+         log.debugf("%s: flush(%s)", poolName, mode);
 
-      Set<ManagedConnectionPool> clearMcpPools = new HashSet<ManagedConnectionPool>();
-      int size = mcpPools.size();
+         Set<ManagedConnectionPool> clearMcpPools = new HashSet<ManagedConnectionPool>();
+         int size = mcpPools.size();
 
-      Iterator<ManagedConnectionPool> it = mcpPools.values().iterator();
-      while (it.hasNext())
-      {
-         ManagedConnectionPool mcp = it.next();
-         try
+         Iterator<ManagedConnectionPool> it = mcpPools.values().iterator();
+         while (it.hasNext())
          {
-            mcp.flush(mode);
-         }
-         catch (Exception e)
-         {
-            // Should not happen
-            log.trace("MCP.flush: " + e.getMessage(), e);
-         }
+            ManagedConnectionPool mcp = it.next();
+            try
+            {
+               mcp.flush(mode, removedConnectionListeners);
+            }
+            catch (Exception e)
+            {
+               // Should not happen
+               log.trace("MCP.flush: " + e.getMessage(), e);
+            }
 
-         if (mcp.isEmpty() && !isPrefill() && size > 1)
-            clearMcpPools.add(mcp);
-      }
+            if (mcp.isEmpty() && !isPrefill() && size > 1)
+               clearMcpPools.add(mcp);
+         }
 
-      if (clearMcpPools.size() > 0)
-      {
-         for (ManagedConnectionPool mcp : clearMcpPools)
+         if (clearMcpPools.size() > 0)
          {
-            if (mcp.isEmpty())
+            for (ManagedConnectionPool mcp : clearMcpPools)
             {
-               try
+               if (mcp.isEmpty())
                {
-                  mcp.shutdown();
-               }
-               catch (Exception e)
-               {
-                  // Should not happen
-                  log.trace("MCP.shutdown: " + e.getMessage(), e);
-               }
+                  try
+                  {
+                     mcp.shutdown();
+                  }
+                  catch (Exception e)
+                  {
+                     // Should not happen
+                     log.trace("MCP.shutdown: " + e.getMessage(), e);
+                  }
 
-               if (Tracer.isEnabled())
-                  Tracer.destroyManagedConnectionPool(getName(), mcp);
-                     
-               mcpPools.values().remove(mcp);
+                  if (Tracer.isEnabled())
+                     Tracer.destroyManagedConnectionPool(getName(), mcp);
+
+                  mcpPools.values().remove(mcp);
+               }
             }
-         }
 
-         log.tracef("%s: mcpPools=%s", getName(), mcpPools);
+            log.tracef("%s: mcpPools=%s", getName(), mcpPools);
+         }
+      }
+      if (removedConnectionListeners != null)
+      {
+         removedConnectionListeners.parallelStream().forEach(cl ->{
```

Comment:
that's ok in 1.4. Just don't forget we can't rely on JDK 1.8 features backporting to 1.3 and 1.0. And since the original case is on 1.0 backporting is definitely needed